kubelet: Make service environment variables optional

vadorovsky opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:
1. If this is your first time, read our contributor guidelines https://git.k8s.io/community/contributors/devel/pull-requests.md#the-pr-submit-process and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/devel/pull-requests.md#best-practices-for-faster-reviews
3. Follow the instructions for writing a release note: https://git.k8s.io/community/contributors/devel/pull-requests.md#write-release-notes-if-needed
4. If the PR is unfinished, see how to mark it: https://github.com/kubernetes/community/blob/master/contributors/devel/pull-requests.md#marking-unfinished-pull-requests
-->

**What this PR does / why we need it**:
Services should be accessed by DNS, not by environment variables which are associated with the deprecated --link option in Docker. In this change, those variables can be optionally enabled by setting up the `kubernetes.io/service-environment-variables` annotation.

**Which issue(s) this PR fixes** *(optional, in `fixes #<issue number>(, fixes #<issue_number>, ...)` format, will close the issue(s) when PR gets merged)*:
Fixes #60099

**Special notes for your reviewer**:

**Release note**:
<!--  Write your release note:
1. Enter your extended release note in the below block. If the PR requires additional action from users switching to the new release, include the string "action required".
2. If no release note is required, just write "NONE".
-->
```release-note
Make service environment variables optional
```

thockin commented:
Please start with this:  https://github.com/kubernetes/community/pull/1249

We have a mostly-agreed design.

thockin commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -557,18 +558,26 @@ func (kl *Kubelet) getServiceEnvVarMap(ns string) (map[string]string, error) {
-	serviceEnv, err := kl.getServiceEnvVarMap(pod.Namespace)
-	if err != nil {
-		return result, err
+	if _, ok := pod.Annotations[kubeapi.ServiceEnvironmentVariablesKey]; ok {
```
This should be a field, not an annotation.

mrostecki replied:
done

jdumars commented:
/ok-to-test

thockin commented:
/hold

ncdc commented:
/uncc

thockin commented:
I very much want to see this happen.  I hope you have time to follow through.  Please assign to me when ready for review.

bradhoekstra commented:
I'm also interested in this feature. It looks like it hasn't been touched in while.

@mrostecki I am willing to help out/take over completion of this PR if you aren't working on it right now.

mrostecki commented:
@bradhoekstra Sorry for the lack of activity. I will rebase and fix this PR today, in regard to the latest discussion on the proposal.

bradhoekstra commented:
Perfect, thanks!

thockin commented:
please assign me when ready.

thockin commented on staging/src/k8s.io/api/core/v1/types.go:
```diff
@@ -2954,6 +2954,10 @@ type PodSpec struct {
 	// configuration based on DNSPolicy.
 	// +optional
 	DNSConfig *PodDNSConfig `json:"dnsConfig,omitempty" protobuf:"bytes,26,opt,name=dnsConfig"`
+	// EnableServiceLinks indicates whether information about services should be injected into pod's
```
document the default

thockin commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -538,6 +538,11 @@ func (kl *Kubelet) getServiceEnvVarMap(ns string) (map[string]string, error) {
 		case ns:
+			// If service links are not enabled, we shouldn't add services from
+			// the same namespace.
+			if !*enableServiceLinks {
```
I think we need to special-case KUBERNETES_* vars (master) because clients need it.

bradhoekstra commented:
@mrostecki Any progress on this?

mrostecki commented:
@bradhoekstra Yes, please expect update on my PR tomorrow. Sorry for making you wait!

vadorovsky committed:
kubelet: Make service environment variables optional

Services should be accessed by DNS, not by environment variables
which are associated with the deprecated --link option in Docker.
Those variables can be optionally enabled by setting up the
kubernetes.io/service-environment-variables annotation.

Fixes #60099

vadorovsky committed:
api: Add EnableServiceLinks field to PodSpec

EnableServiceLinks field indicates whether information about
services should be injected into pod's environment variables,
matching the syntax of Docker links.

thockin commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -536,6 +536,11 @@ func (kl *Kubelet) getServiceEnvVarMap(ns string) (map[string]string, error) {
 	for _, e := range envvars.FromServices(mappedServices) {
+		// If service links are not enabled, we shouldn't add service variables
+		// except those which start from "KUBERNETES_".
+		if !*enableServiceLinks && !strings.HasPrefix(e.Name, "KUBERNETES_") {
```
It's rather wasteful to call FromServices and then throw away most of it, especially if we want this to be the default mode eventually.

Can we instead change the loop above to test the flag before adding services to the `serviceMap` ?

thockin commented on pkg/kubelet/kubelet_pods.go:
```diff
@@ -544,6 +549,8 @@ func (kl *Kubelet) getServiceEnvVarMap(ns string) (map[string]string, error) {
 // Make the environment variables for a pod in the given namespace.
 func (kl *Kubelet) makeEnvironmentVariables(pod *v1.Pod, container *v1.Container, podIP string) ([]kubecontainer.EnvVar, error) {
 	var result []kubecontainer.EnvVar
+	var serviceEnv map[string]string
```
Why this change?  In general, scoping closer to usage is better.

thockin commented on pkg/kubelet/kubelet_pods_test.go:
```diff
@@ -834,20 +834,22 @@ func TestMakeEnvironmentVariables(t *testing.T) {
-			ns:   "test1",
+			name:               "api server = Y, kubelet = Y",
+			ns:                 "test1",
+			enableServiceLinks: false,
```
It's not clear what these test cases think they are testing, but you're changing it.  Since the default is true, you should probably handle this as:

1) add `enableServiceLinks: true` to each test case
2) add new test cases (with comments, please) that explicitly exercise the `false` case of that and nothing else.

If you feel generous you could refactor this test, making a util function that takes a list of test cases and a bunch of smaller Test* functions that call that util function with more targetted cases, e.g `TestMakeEnvironmentVariablesWithExpansions` and `TestMakeEnvironmentVariablesFromConfigMap` etc.  You can do that a a pre-factoring commit in this same PR.  If you feel generous...  It would make this testing much nicer :)

mrostecki commented:
@bradhoekstra OK, we can close this one in favor of yours.
